projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+// standard includes
+#include<string>
+#include<vector>
+#include <thread>
+#include <mutex>
+#include <atomic>
+
+// BoB includes
+#include "common/main.h"
+#include "common/stopwatch.h"
+#include "common/background_exception_catcher.h"
+#include "common/gps.h"
+#include "common/map_coordinate.h"
+#include "common/bn055_imu.h"
+
+#include "imgproc/opencv_unwrap_360.h"
+#include "video/panoramic.h"
+
+#include <sys/stat.h>
+#include <ctime>
+
+
+
+
+
+
+// Standard C++ includes
+#include <iostream>
+
+using namespace BoBRobotics;
+using namespace BoBRobotics::ImgProc;
+using namespace BoBRobotics::Video;
+
+
+#define WIDTH 192
+#define HEIGHT 72
+
+#define IS_UNWRAP true
+#define UPDATE_INTERVAL 200
+
+
+void readCameraThreadFunc(BoBRobotics::Video::Input *cam,
+                        cv::Mat &img,
+                        std::atomic<bool> &shouldQuit,
+                        std::mutex &mutex)
+{
+
+    //auto unwrapper = cam->createUnwrapper(cv::Size(WIDTH,HEIGHT));
+
+    while(!shouldQuit) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(5));
+        std::lock_guard<std::mutex> lock(mutex);
+        cam->readFrame(img);
+
+    }
```

Comment:
I think you probably already know my opinion on this threading business by now :wink:. I feel we probably could remove it, but as this isn't a priority we can look at this later.

Longer term I think we want to establish whether or not it actually fixes camera problems and if it does, we should do this threading malarkey in ``Video::OpenCVCamera`` and if it isn't we should drop it.

Reply:
Would you mind removing this thread too? I'm pretty sure the implementation is correct, but we still don't need it and removing it will make the code cleaner. The fewer threads we have, the better.

If you're doing it so that you can initialise the hardware in parallel, use ``std::async`` (see here for example: https://github.com/BrainsOnBoard/bob_robotics/blob/master/projects/snapshot_bot/snapshot_bot.cc)